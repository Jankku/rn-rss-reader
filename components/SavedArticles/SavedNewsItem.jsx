import { memo } from 'react';
import { StyleSheet } from 'react-native';
import { useTheme } from '@react-navigation/native';
import { Dimensions, Image, Pressable, Text, View } from 'react-native';
import { stringOrPlaceholder } from '../../utils/stringutils';
import { RFC2822ToDateTime } from '../../utils/dateutils';

function SavedNewsItem({ title, description, imageUrl, pubDate, onPress }) {
  const { colors, dark } = useTheme();
  const styles = makeStyles(colors, dark);

  return (
    <Pressable
      onPress={onPress}
      style={({ pressed }) => ({
        backgroundColor: pressed ? colors.background : colors.card,
      })}
    >
      <View style={styles.container}>
        <View style={styles.textContainer}>
          <Text style={styles.title}>{stringOrPlaceholder(title)}</Text>
          <Text style={styles.description}>{stringOrPlaceholder(description)}</Text>
          <Text style={styles.date}>{stringOrPlaceholder(RFC2822ToDateTime(pubDate))}</Text>
        </View>
        <View style={styles.imageContainer}>
          <Image resizeMode="cover" style={styles.image} source={{ uri: imageUrl }} />
        </View>
      </View>
    </Pressable>
  );
}

const makeStyles = (colors, dark) =>
  StyleSheet.create({
    container: {
      flex: 1,
      flexDirection: 'row',
      paddingHorizontal: 12,
      paddingVertical: 16,
      width: Dimensions.get('window').width,
    },
    date: { color: colors.text, fontSize: 12 },
    description: { color: colors.text, paddingBottom: 6 },
    image: {
      backgroundColor: `rgba(${dark ? '255, 255, 255' : '0, 0, 0'}, 0.1)`,
      borderRadius: 8,
      height: 100,
      marginLeft: 4,
      marginTop: 6,
      width: 100,
    },
    imageContainer: { paddingLeft: 4 },
    textContainer: { flex: 1 },
    title: { color: colors.text, fontWeight: '700', paddingBottom: 4 },
  });

export default memo(SavedNewsItem);
