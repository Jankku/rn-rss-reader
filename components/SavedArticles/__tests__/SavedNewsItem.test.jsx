import SavedNewsItem from '../SavedNewsItem';
import { render, fireEvent, waitFor } from '@testing-library/react-native';

describe('Test SavedNewsItem', () => {
  test('Should match snapshot', () => {
    const { toJSON } = render(
      <SavedNewsItem
        title={'title'}
        description={'description'}
        imageUrl={'imageUrl'}
        pubDate={'04/04/2022, 19:00'}
        onPress={() => {}}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('Should call onPress', () => {
    const onPress = jest.fn();
    const { getByText } = render(
      <SavedNewsItem
        title={'title'}
        description={'description'}
        imageUrl={'imageUrl'}
        pubDate={'04/04/2022, 19:00'}
        onPress={onPress}
      />
    );
    const item = getByText('title');

    fireEvent.press(item);
    expect(onPress).toHaveBeenCalledTimes(1);
  });

  test('Should use placeholder values', async () => {
    const { getAllByText } = render(
      <SavedNewsItem
        title={''}
        description={''}
        imageUrl={'imageUrl'}
        pubDate={''}
        onPress={() => {}}
      />
    );
    const textElements = getAllByText('-');
    await waitFor(() => expect(textElements).toHaveLength(3));
  });
});
