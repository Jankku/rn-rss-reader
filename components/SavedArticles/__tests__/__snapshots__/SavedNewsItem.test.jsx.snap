// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SavedNewsItem Should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgb(255, 255, 255)",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "paddingHorizontal": 12,
        "paddingVertical": 16,
        "width": 750,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
            "fontWeight": "700",
            "paddingBottom": 4,
          }
        }
      >
        title
      </Text>
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
            "paddingBottom": 6,
          }
        }
      >
        description
      </Text>
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
            "fontSize": 12,
          }
        }
      >
        04/04/2022, 19:00
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 4,
        }
      }
    >
      <Image
        resizeMode="cover"
        source={
          Object {
            "uri": "imageUrl",
          }
        }
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.1)",
            "borderRadius": 8,
            "height": 100,
            "marginLeft": 4,
            "marginTop": 6,
            "width": 100,
          }
        }
      />
    </View>
  </View>
</View>
`;
